### PR TITLE
Support patterns with nested colons in case statements

### PR DESCRIPTION
**Pull request by 333fred**
The case statement parser looked until it found the first `:`. This isn't quite right: patterns can have nested `:` tokens. Instead, it needs to parse out recursive, tuple, and list patterns until it finds a `:` at the top level, and then it can consider the case clause complete. Fixes https://github.com/dotnet/razor-compiler/issues/220.

### CODE REVIEW

**Comment by 333fred** [comment]
@RikkiGibson @chsienki for review.

**Comment by 333fred** [comment]
@RikkiGibson @chsienki for reviews please.

File: src/test/Microsoft.AspNetCore.Razor.Test.Common/Language/IntegrationTests/InitializeTestFileAttribute.cs [hunk 1]
```diff
@@ -8,7 +8,7 @@
 
 namespace Microsoft.AspNetCore.Razor.Language.IntegrationTests;
 
-public class IntializeTestFileAttribute : BeforeAfterTestAttribute
+public class InitializeTestFile : BeforeAfterTestAttribute
```
Comment (chsienki) [inline src/test/Microsoft.AspNetCore.Razor.Test.Common/Language/IntegrationTests/InitializeTestFileAttribute.cs:11]: Nit: keep the attribute postfix for the actual type name?

**Review by chsienki — APPROVED** [review]

**Comment by 333fred** [comment]
@RikkiGibson @dotnet/razor-compiler for a second review.

**Review by RikkiGibson — APPROVED** [review]

**Pull request merged**